src/lib/blkalloc/bitmap_blk_allocator.cpp

```diff
@@ -0,0 +1,201 @@
+/*********************************************************************************
+ * Modifications Copyright 2017-2019 eBay Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ *********************************************************************************/
+#include <homestore/homestore.hpp>
+#include <homestore/meta_service.hpp>
+#include <homestore/checkpoint/cp_mgr.hpp>
+#include "bitmap_blk_allocator.h"
+#include "meta/meta_sb.hpp"
+#include "common/homestore_utils.hpp"
+
+namespace homestore {
+BitmapBlkAllocator::BitmapBlkAllocator(BlkAllocConfig const& cfg, bool is_fresh, chunk_num_t id) :
+        BlkAllocator(cfg, id), m_blks_per_portion{cfg.m_blks_per_portion} {
+    if (is_persistent()) {
+        meta_service().register_handler(
+            get_name(),
+            [this](meta_blk* mblk, sisl::byte_view buf, size_t size) {
+                on_meta_blk_found(voidptr_cast(mblk), std::move(buf), size);
+            },
+            nullptr);
+    }
+
+    if (is_fresh || !is_persistent()) {
+        m_disk_bm = std::make_unique< sisl::Bitset >(m_num_blks, m_chunk_id, m_align_size);
+        do_init();
+    }
+}
+
+void BitmapBlkAllocator::do_init() {
+    // NOTE:  Blocks per portion must be modulo word size so locks do not fall on same word
+    m_blks_per_portion = sisl::round_up(m_blks_per_portion, m_disk_bm ? m_disk_bm->word_size() : 64u);
+
+    m_blk_portions = std::make_unique< BlkAllocPortion[] >(get_num_portions());
+    for (blk_num_t index{0}; index < get_num_portions(); ++index) {
+        m_blk_portions[index].set_portion_num(index);
```

Comment:
We have removed the set_available_nblks() altogether, because after restart it is not consistent and we don't load the correct number of blks available. So it doesn't make sense to do that.